radicle-seed/src/lib.rs

```diff
@@ -0,0 +1,271 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![allow(clippy::single_match)]
+#[macro_use]
+extern crate async_trait;
+
+use futures::stream::StreamExt;
+use std::{collections::HashSet, io, net::SocketAddr, path::PathBuf, vec};
+use thiserror::Error;
+
+use radicle_keystore::sign::ed25519;
+
+use librad::{
+    git,
+    keys,
+    meta::project,
+    net::{
+        discovery,
+        gossip,
+        gossip::types::PeerInfo,
+        peer::{self, PeerApi, PeerConfig},
+        protocol::ProtocolEvent,
+    },
+    paths,
+    peer::PeerId,
+    uri::RadUrn,
+};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Storage(#[from] git::storage::Error),
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+
+    #[error(transparent)]
+    Bootstrap(#[from] peer::BootstrapError),
+
+    #[error(transparent)]
+    Accept(#[from] peer::AcceptError),
+}
+
+#[derive(Clone)]
+pub struct Signer {
+    key: keys::SecretKey,
+}
+
+impl Signer {
+    pub fn new<R: io::Read>(mut r: R) -> Result<Self, io::Error> {
+        use radicle_keystore::SecretKeyExt;
+
+        let mut bytes = Vec::new();
+        r.read_to_end(&mut bytes)?;
+
+        let sbytes: keys::SecStr = bytes.into();
+        match keys::SecretKey::from_bytes_and_meta(sbytes, &()) {
+            Ok(key) => Ok(Self { key }),
+            Err(err) => Err(io::Error::new(io::ErrorKind::InvalidData, err)),
+        }
+    }
+}
+
+#[async_trait]
+impl ed25519::Signer for Signer {
+    type Error = std::convert::Infallible;
+
+    fn public_key(&self) -> ed25519::PublicKey {
+        self.key.public_key()
+    }
+
+    async fn sign(&self, data: &[u8]) -> Result<ed25519::Signature, Self::Error> {
+        <keys::SecretKey as ed25519::Signer>::sign(&self.key, data).await
+    }
+}
+
+impl keys::AsPKCS8 for Signer {
+    fn as_pkcs8(&self) -> Vec<u8> {
+        self.key.as_pkcs8()
+    }
+}
+
+/// Seed operational mode.
+#[derive(Debug)]
+pub enum Mode {
```

Comment:
Minor nit, but instead of prefixing `Track` to every variant you coul rename it to `TrackMode`.

Reply:
True, I think we will eventually change this anyway to `TrackRule` and have them be a Monoid of sorts